### DISCUSSION TITLE
Enable Cell Picking Changes Scalar Visualization

### DISCUSSION THREAD

**Discussion by teliason** [pyvista/pyvista#2405] [Q&A] [answered]
I am trying to create a tool that allows the user to visually investigate data mapped to a mesh by hiding elements that the user selects.  Similar to how you would look at fringe plots within various FEA post processors.  When not using any scalar data the code I have works exactly as I intend, however when scalar data is assigned the enable_cell_picking function seems to change or remove the visualization of the scalar assignments.  

Plotter before selection is made
![image](https://user-images.githubusercontent.com/102766653/161152271-b4a9f9cd-c30c-4410-8764-7c1386c916e9.png)

Plotter after selection is made
![image](https://user-images.githubusercontent.com/102766653/161152399-e25a15fe-b361-464c-9eb6-1b4152cd16f9.png)

Is the selection process in enable_cell_picking changing something to how the scalar values are assigned or viewed?  Or do I need to do something after that selection is made in the callback function to update the scalar values again?

Code used to make the example images.

```py
import numpy as np
import vtk
import pyvista as pv
from numpy import random
  
mesh = pv.Sphere()   
  
def update_plot(selection):
    
    if not selection.n_cells:
        return
    
    ghost_cells = np.zeros(pl.mesh.n_cells, np.uint8)
    ghost_cells[selection['orig_extract_id']] = 1
    pl.mesh.cell_data[vtk.vtkDataSetAttributes.GhostArrayName()] = ghost_cells
    pl.mesh.RemoveGhostCells()
    
def reset_plot():
    
    mesh = orig_mesh.copy()
    pl.clear()
    pl.add_mesh(mesh, show_edges=True)
    pl.enable_cell_picking(mesh=mesh, callback=update_plot, show=False)
       
values = random.rand(mesh.number_of_cells)

mesh.cell_data['values'] = values

orig_mesh = mesh.copy()  

pl = pv.Plotter()
pl.add_mesh(mesh, show_edges=True)
pl.enable_cell_picking(mesh=mesh, callback=update_plot, show=False)
pl.add_key_event('o', reset_plot)
pl.show()
```

**Comment by akaszynski** [comment, marked as answer]
Cool demo!

Issue is that the `enable_cell_picking` callback changes the active cell scalars. You can override that by setting the active scalars again with `pl.mesh.cell_data.active_scalars_name = 'values'`.

Now, it works:

https://user-images.githubusercontent.com/11981631/161272442-243236a9-5712-4759-b3d5-6910900757d1.mp4



#### Code

```py
import numpy as np
import vtk
import pyvista as pv
from numpy import random

pv.set_plot_theme('document')

mesh = pv.Sphere()

def update_plot(selection):
    if not selection.n_cells:
        return

    ghost_cells = np.zeros(pl.mesh.n_cells, np.uint8)
    ghost_cells[selection['orig_extract_id']] = 1
    pl.mesh.cell_data[vtk.vtkDataSetAttributes.GhostArrayName()] = ghost_cells
    pl.mesh.RemoveGhostCells()
    pl.mesh.cell_data.active_scalars_name = 'values'  # added

def reset_plot():
    mesh = orig_mesh.copy()
    pl.clear()
    pl.add_mesh(mesh, show_edges=True)
    pl.enable_cell_picking(mesh=mesh, callback=update_plot, show=False)

values = random.rand(mesh.number_of_cells)

mesh.cell_data['values'] = values

orig_mesh = mesh.copy()

pl = pv.Plotter()
pl.add_mesh(mesh, show_edges=True)
pl.enable_cell_picking(mesh=mesh, callback=update_plot, show=False)
pl.add_key_event('o', reset_plot)
pl.show()
```

**Reply by teliason** [reply]
Awesome, I figured it was something simple that I was missing.  That's exactly what I needed, thanks for the help I really appreciate it.
